### ISSUE TITLE
match_fginn: geometric consistency check indexes the wrong axis, so FGINN degenerates to match_snn

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4062]
### Describe the bug

`match_fginn` indexes the wrong axis when it computes the spatial distance between a query's 1st nearest neighbour and its remaining candidates. As a result the "first geometrically inconsistent nearest neighbour" test — the entire point of FGINN (:cite:`MODS2015`) — never fires, and `match_fginn` degenerates into plain `match_snn` for essentially every query.

https://github.com/kornia/kornia/blob/main/kornia/feature/matching.py#L303-L305

```python
candidates_xy = xy2[idxs_in_2]                                       # (B1, num_candidates, 2)
kdist = torch.norm(candidates_xy - candidates_xy[0:1], p=2, dim=2)   # <-- indexes query 0
fginn_vals = vals_cand[:, 1:] + (kdist[:, 1:] < spatial_th).to(dtype) * BIG_NUMBER
```

`candidates_xy` has shape `(B1, num_candidates, 2)`, so `candidates_xy[0:1]` selects **query 0's entire candidate list**, shape `(1, num_candidates, 2)`, and broadcasts it across every query. That makes

```
kdist[i, k] = || xy2[idx[i, k]] - xy2[idx[0, k]] ||
```

i.e. query *i*'s k-th candidate measured against **query 0**'s k-th candidate, instead of the intended

```
kdist[i, k] = || xy2[idx[i, k]] - xy2[idx[i, 0]] ||
```

The slice should be `candidates_xy[:, 0:1]`.

### Consequences

Three distinct symptoms, all reproduced below:

1. **The geometric term is inert.** For an arbitrary query, its candidates are compared against an unrelated query's candidates, which are almost always farther apart than `spatial_th`, so nothing is penalised and the raw 2nd NN is used — exactly what `match_snn` does.
2. **Query 0 always matches.** For `i == 0` the expression is `candidates_xy[0] - candidates_xy[0] ≡ 0`, so *every* candidate is flagged as spatially close and gets `+BIG_NUMBER`. The 2nd-NN distance becomes enormous and the ratio collapses to ~0, so the first query passes any threshold.
3. **Cross-query contamination.** A query's returned ratio depends on an unrelated query in the same batch.

### Reproduction

Ratios are bit-identical to `match_snn`:

```python
import torch
import kornia.feature.matching as M
from kornia.feature.laf import laf_from_center_scale_ori

torch.manual_seed(0)
D, N1, N2 = 128, 300, 300
desc1 = torch.nn.functional.normalize(torch.randn(N1, D), dim=1)
desc2 = torch.nn.functional.normalize(torch.randn(N2, D), dim=1)
xy2 = torch.rand(N2, 2) * 640
for k in range(0, N2, 2):                       # planted near-duplicates a few px apart
    desc2[k + 1] = torch.nn.functional.normalize(desc2[k] + 0.15 * torch.randn(D), dim=0)
    xy2[k + 1] = xy2[k] + torch.randn(2) * 3.0
lafs2 = laf_from_center_scale_ori(xy2[None], torch.ones(1, N2, 1, 1))
lafs1 = laf_from_center_scale_ori(torch.rand(1, N1, 2) * 640, torch.ones(1, N1, 1, 1))

d_snn, i_snn = M.match_snn(desc1, desc2, 0.95)
d_fg, i_fg = M.match_fginn(desc1, desc2, lafs1, lafs2, th=0.95, spatial_th=10.0)

m_snn = {tuple(r): v for r, v in zip(i_snn.tolist(), d_snn.view(-1).tolist())}
m_fg = {tuple(r): v for r, v in zip(i_fg.tolist(), d_fg.view(-1).tolist())}
common = set(m_snn) & set(m_fg)
same = sum(1 for k in common if abs(m_snn[k] - m_fg[k]) < 1e-6)
print(f"of {len(common)} matches shared with match_snn, {same} have an IDENTICAL ratio")
```

```
of 15 matches shared with match_snn, 15 have an IDENTICAL ratio
```

With `candidates_xy[:, 0:1]` this drops to 14/15 — the spatial suppression actually fires.

Cross-query contamination, same query in every run:

```python
import torch
from kornia.feature.matching import match_fginn
from kornia.feature.laf import laf_from_center_scale_ori

torch.manual_seed(0)
D, N2 = 8, 12
desc2 = torch.nn.functional.normalize(torch.randn(N2, D), dim=1)
desc2[4] = torch.nn.functional.normalize(desc2[3] + 0.02 * torch.randn(D), dim=0)
xy2 = torch.rand(N2, 2) * 200.0
xy2[4] = xy2[3] + torch.tensor([2.0, 0.0])       # duplicate 2px from the 1st NN
lafs2 = laf_from_center_scale_ori(xy2[None], torch.ones(1, N2, 1, 1))
q = torch.nn.functional.normalize(desc2[3] + 0.01 * torch.randn(D), dim=0)

def ratio_for_q(first_query):
    desc1 = torch.stack([first_query, q])
    lafs1 = laf_from_center_scale_ori(torch.zeros(1, 2, 2), torch.ones(1, 2, 1, 1))
    d, i = match_fginn(desc1, desc2, lafs1, lafs2, th=1.0, spatial_th=10.0)
    hit = d[i[:, 0] == 1]
    return hit[0, 0].item() if len(hit) else None

torch.manual_seed(11); print(ratio_for_q(torch.nn.functional.normalize(desc2[7] + 0.05 * torch.randn(D), dim=0)))
torch.manual_seed(14); print(ratio_for_q(q.clone()))
```

```
0.9005319476127625
2.797479226046562e-08
```

Query #1's descriptor and LAFs are identical in both runs; only the unrelated query #0 changed.

### Expected behavior

`kdist[i, k]` should measure candidate *k* of query *i* against the 1st NN **of that same query**, so that candidates lying within `spatial_th` of the 1st NN are excluded from the ratio test. Each query's result must be independent of the other queries in the batch.

### Additional context

- `lafs1` is accepted and documented by `match_fginn` but never read — only `lafs2` is used.
- Fixing this changes every `match_fginn` / `DescriptorMatcher("fginn")` / `GeometryAwareDescriptorMatcher("fginn")` result, so it needs a deliberate call rather than a silent patch.

Found during an audit of `kornia.feature`.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
